using Distributions, Random, LinearAlgebra, Parameters, DataFrames, CSV, Documenter, StatsBase, Plots

"""
model parameters
- `ğ”¼P` = expected total phenotypic variance
- `hÂ²` = expected proportion of additive genetic variance
- `L` = num host gen loci (set to zero for no host genetics)
- `SL` = num lin trans microbes (set to zero for no lin micr's)
- `SN` = num non-lin trans microbes (set to zero for no non-lin micr's)
- `SE` = num env acqr microbes (set to zero for no env micr's)
"""
@with_kw mutable struct Par

    # expected total phenotypic variance
    ğ”¼P::Float64 = 100.0

    # expected proportion of additive genetic variance
    hÂ²::Float64 = 0.2

    # expected proportion of additive lineal microbial variance
    Î»Â²::Float64 = 0.2

    # expected proportion of additive non-lineal vertical microbial variance
    Î½Â²::Float64 = 0.2

    # expected proportion of additive external microbial variance
    ÏµÂ²::Float64 = 0.2

    # probability of mutant allele
    p::Float64 = 0.5

    # number of genetic loci
    L::Int64 = 100

    # number of lineally transmitted taxa
    SL::Int64 = 100

    # number of non-lineally vertically transmitted taxa
    SN::Int64 = 100

    # number of uninheritable environmentally acquired taxa
    SE::Int64 = 100

    # number of hosts in each generation
    n::Int64 = 10^3

    # strength of directional selection
    s::Float64 = 1e-1

    # probability of lineal transmission per lineal microbe (given presence of that microbe)
    Î»::Float64 = 0.25 / SL

    # probability of non-lineal vert transmission per non-lineal microbe (given presence of that microbe)
    Î½::Float64 = 0.25 / (SN*n)

    #
    # numbers of times to repeat things
    #

    # number of times to initialize parental population
    nr::Int64 = 30
    
    # number of times to repeat selection experiments
    sr::Int64 = 5
    
    # number of times to repeat production of offspring generation
    or::Int64 = 5

    # total number of simulation runs = nr * sr * or

    # expected abundance of each microbial taxa in each host parent
    KL = 100
    KN = 100
    KE = 100

    # whether nonlineal microbes are sourced from selected parents
    Nsel = true

end

"""
parameters for generating population data (`PP` = `P`opulation `P`arameters)
- determines trait architecture shared by parents and offspring
- `Î³` = additive genetic fx
- `Ï‰L` = additive lineal microbial fx
- `Ï‰N` = additive non-lineal microbial fx
- `Ï‰E` = additive environmental microbial fx
"""
@with_kw mutable struct PP
    Î³::Vector{Float64}
    Ï‰L::Vector{Float64}
    Ï‰N::Vector{Float64}
    Ï‰E::Vector{Float64}
end

"""
population data (`PD` = `P`opulation `D`ata)
- `g` = host (diploid) genotype array
- `mL` = lineal microbe abundance matrix
- `mN` = non-lineal microbe abundance matrix
- `mE` = external microbe abundance matrix
- `z` = host trait vector
"""
@with_kw mutable struct PD
    g::Array{Bool,3}
    mL::Matrix{Int64}
    mN::Matrix{Int64}
    mE::Matrix{Int64}
    z::Vector{Float64}
end

"""
initialize parents
- returns `PP` and `PD`
"""
function init(par::Par)
    @unpack_Par par

    # expected additive genetic variance
    ğ”¼GA = hÂ²*ğ”¼P

    # expected additive lineal microbial variance
    ğ”¼ML = Î»Â²*ğ”¼P

    # expected additive non-lineal microbial variance
    ğ”¼MN = Î½Â²*ğ”¼P

    # expected additive environmental microbial variance
    ğ”¼ME = ÏµÂ²*ğ”¼P

    # expected developmental noise / environmental deviation
    ğ”¼E = (1 - hÂ² - Î»Â² - Î½Â² - ÏµÂ²) * ğ”¼P

    # total number of microbe taxa across categories
    S = SL + SN + SE

    # unscaled variance of additive genetic fx
    vÎ“ = ğ”¼GA / (2*p*(1-p))

    # unscaled variance of additive lineal microbial fx
    vL = ğ”¼ML / (KL)

    # unscaled variance of additive non-lineal microbial fx
    vN = ğ”¼MN / (KN)

    # unscaled variance of additive non-lineal microbial fx
    vE = ğ”¼ME / (KE)

    #
    # parental parameters
    #

    # additive fx of genetic loci
    if L > 0
        Î³ = rand(Normal(0,âˆš(vÎ“/L)),L)
    else
        Î³ = 0
    end

    # additive fx of lineal microbes
    if SL > 0
        Ï‰L = rand(Normal(0,âˆš(vL/SL)),SL)
    else
        Ï‰L = 0
    end

    # additive fx of non-lineal heritable microbes
    if SN > 0
        Ï‰N = rand(Normal(0,1âˆš(vN/SN)),SN)
    else
        Ï‰N = 0
    end

    # additive fx of environmental microbes
    if SE > 0
        Ï‰E = rand(Normal(0,âˆš(vE/SE)),SE)
    else
        Ï‰E = 0
    end

    #
    # parental data
    #

    # genetic array
    g = rand(Bernoulli(p),n,L,2)

    # lineal microbiome matrix
    mL = rand(Poisson(KL),n,SL)

    # nonlineal microbiome matrix
    mN = rand(Poisson(KN),n,SN)

    # environmental microbiome matrix
    mE = rand(Poisson(KE),n,SE)

    #
    # collecting parental parameters and data
    #

    # collect parental parameters
    pp = PP(
        Î³ = Î³,
        Ï‰L = Ï‰L,
        Ï‰N = Ï‰N,
        Ï‰E = Ï‰E
    )

    # collect parental data
    pd = PD(
        g = g,
        mL = mL,
        mN = mN,
        mE = mE,
        z = zeros(n)
    )

    #
    # traits
    #

    # genetic-microbic values
    zgm = gmvals(pp,pd,"Î³Î»Î½Îµ")

    # (generative) developmental noise / environmental deviation
    e = rand(Normal(0,âˆšğ”¼E),n)

    # compute parental trait values from (generative) model
    pd.z = zgm .+ e

    return pp, pd
end

"""
compute genetic-microbic value of host traits
- used during formation of host parents and host offspring data
- `e` will need to be added later if noise is desired
"""
function gmvals(pp::PP,pd::PD)
    @unpack_PP pp
    @unpack_PD pd

    # allele count matrix
    c = g[:,:,1] .+ g[:,:,2]

    # (generative) additive genetic value
    gA = c*Î³

    # (generative) lineal microbial value
    ml = mL*Ï‰L

    # (generative) nonlineal microbial value
    mn = mN*Ï‰N

    # (generative) environmental microbial value
    me = mE*Ï‰E

    # (generative) additive microbial value
    mA = ml .+ mn .+ me

    # compute parental trait values from (generative) model
    zgm = gA .+ mA

    return zgm
end

"""
compute genetic-microbic value of host traits
- used for the analysis of variance
- just uses model of `zgm` instead of actually averaging
- `fs` stands for *factors*
    - `fs = "Î³Î»Î½Îµ"` for estimating zgm from all microbes
    - `fs = "Î³Î»Î½"` for estimating zgm excluding env microbes
    - `fs = "Î³Î»Î½"` for estimating zgm including only lineal microbes
    - `fs = "Î³"` for estimating zgm from only genetic data
"""
function gmvals(pp::PP,pd::PD,fs::String)
    @unpack_PP pp
    @unpack_PD pd

    # allele count matrix
    c = g[:,:,1] .+ g[:,:,2]

    # (mechanistic) additive genetic value
    gA = c*Î³

    if fs == "Î³Î»Î½Îµ"

        # (mechanistic) additive lineal microbial value
        ml = mL*Ï‰L
        
        # (mechanistic) additive nonlineal microbial value
        mn = mN*Ï‰N

        # (mechanistic) additive environmental microbial value
        me = mE*Ï‰E

    elseif fs == "Î³Î»Î½"
    
        # (mechanistic) additive lineal microbial value
        ml = mL*Ï‰L
    
        # (mechanistic) additive nonlineal microbial value
        mn = mN*Ï‰N

        # (mechanistic) additive environmental microbial value
        me = 0

    elseif fs == "Î³Î»"
        
        # (mechanistic) additive lineal microbial value
        ml = mL*Ï‰L
    
        # (mechanistic) additive nonlineal microbial value
        mn = 0

        # (mechanistic) additive environmental microbial value
        me = 0

    elseif fs == "Î³"
            
        # (mechanistic) additive lineal microbial value
        ml = 0

        # (mechanistic) additive nonlineal microbial value
        mn = 0

        # (mechanistic) additive environmental microbial value
        me = 0

    else

        print("invalid factor string")

        return 

    end

    # (mechanistic) additive microbial value
    mA = ml .+ mn .+ me

    # compute parental trait values from (generative) model
    zgm = gA .+ mA
    
    return zgm
end

"""
All Additive FX
- includes genetic, lineal, non-lineal, and ext
"""
function AAFX(pp::PP,pd::PD)
    @unpack_PP pp
    @unpack_PD pd

    # compute genetic-microbic values
    zgm = gmvals(pp,pd,"Î³Î»Î½Îµ")
 
    # allele count matrix
    c = g[:,:,1] .+ g[:,:,2]

    # microbiome matrix
    m = [mL mN mE]

    # all the thingis
    a = [c m]

    # gene-microbe covariance matrix
    Î£ = cov(a)

    # additive fx
    Î± = inv(Î£)*cov(a,zgm)

    return Î±[:,1]

end

"""
All Additive Variance
"""
function AA(pp::PP,pd::PD)
    @unpack_PP pp
    @unpack_PD pd

    Î± = AAFX(pp,pd)

    c = g[:,:,1] .+ g[:,:,2]

    a = [c mL mN mE]

    Aâ‚ = var(a*Î±)

    return Aâ‚
end

"""
Genetic and Lineal Additive FX
"""
function LAFX(pp::PP,pd::PD)
    @unpack_PP pp
    @unpack_PD pd

    # compute genetic-microbic values
    zgm = gmvals(pp,pd,"Î³Î»")
 
    # allele count matrix
    c = g[:,:,1] .+ g[:,:,2]

    # genes and lineal microbes
    a = [c mL]

    # gene-microbe covariance matrix
    Î£ = cov(a)

    # additive fx
    Î± = inv(Î£)*cov(a,zgm)

    return Î±[:,1]

end

"""
Genetic and Lineal Additive Variance
"""
function LA(pp::PP,pd::PD)
    @unpack_PP pp
    @unpack_PD pd

    Î± = LAFX(pp,pd)
    
    c = g[:,:,1] .+ g[:,:,2]

    a = [c mL]

    Lâ‚ = var(a*Î±)

    return Lâ‚
end

"""
Additive FX of Transmitted Materials
"""
function TAFX(pp::PP,pd::PD)
    @unpack_PP pp
    @unpack_PD pd

    # compute genetic-microbic values
    zgm = gmvals(pp,pd,"Î³Î»Î½")

    # allele count matrix
    c = g[:,:,1] .+ g[:,:,2]

    # genes and lineal microbes
    a = [c mL mN]

    # gene-microbe covariance matrix
    Î£ = cov(a)

    # additive fx
    Î± = inv(Î£)*cov(a,zgm)

    return Î±[:,1]

end

"""
Additive Variance of Transmitted Materials
"""
function TA(pp::PP,pd::PD)
    @unpack_PP pp
    @unpack_PD pd

    Î± = TAFX(pp,pd)

    c = g[:,:,1] .+ g[:,:,2]

    a = [c mL mN]

    Tâ‚ = var(a*Î±)

    return Tâ‚
end

"""
Additive Genetic FX
"""
function AGFX(pp::PP,pd::PD)
    @unpack_PP pp
    @unpack_PD pd

    # compute genetic-microbic values
    zgm = gmvals(pp,pd,"Î³")

    # allele count matrix
    c = g[:,:,1] .+ g[:,:,2]

    # genetic covariance matrix
    Î£ = cov(c)

    # additive fx for each locus
    Î³ = inv(Î£)*cov(c,zgm)

    return Î³[:,1]

end

"""
Additive Genetic Variance
"""
function GA(pp::PP,pd::PD)
    @unpack_PP pp
    @unpack_PD pd

    Î³ = AGFX(pp,pd)

    c = g[:,:,1] .+ g[:,:,2]

    Gâ‚ = var(c*Î³)

    return Gâ‚
end

"""
selection experiment
- returns indices of selected pairs and num unique pairs
"""
function selection(par::Par, pd::PD)
    @unpack_Par par
    @unpack_PD pd

    # probability of being a member of a parental pair chosen by an offspring
    # proportional to fitness (both types of relative fitness, and absolute fitness)
    w = normalize(exp.(s*z),1)

    # select mating pairs
    wv = Weights(w)
    pinds = collect(1:n)
    p1 = sample(pinds,wv,n)
    p2 = sample(pinds,wv,n)
    pairwents = [p1 p2]
    
    # count number of times each individual in parent population occurs in a mating pair
    W = zeros(Int64, n)
    for i in 1:n
        W[i] += length(findall(x->x==i,pairwents[:,1]))
        W[i] += length(findall(x->x==i,pairwents[:,2]))
    end

    # realized absolute fitness
    W /= 2

    # expected absolute fitness
    ğ”¼W = n*w

    return pairwents, W, ğ”¼W
end

"""
generate offspring traits
- includes transmission of genes and microbes
"""
function offspring(par::Par, pp::PP, pd::PD, pairs)
    @unpack_Par par
    @unpack_PP pp
    @unpack_PD pd

    # whether nonlineal microbes are sourced from selected parents
    if Nsel
        # indices of selected parents        
        vp = vec(pairs)
    else
        # indices of all parents
        vp = 1:n
    end

    # pick nonlineal donors
    nld = sample(vp,n)

    # non-lineal vert trans
    mNâ‚š = rand.(Poisson.(mN[nld,:]))

    # transmission of genetic material and lineal microbes
    gâ‚š = Array{Bool}(undef,n,L,2)
    mLâ‚š = Array{Int64}(undef,n,SL)
    for i in 1:n
        
        p1 = pairs[i,1]
        p2 = pairs[i,2]

        #
        # inheritance of genetic material
        #

        # decide which loci recombine
        free_recomb = rand(Bernoulli(),L)

        # pick one chromosome from each parent
        p1ch = rand(Bernoulli()) .+ 1
        p2ch = rand(Bernoulli()) .+ 1

        # form offspring chromosomes
        ch1 = Array{Bool}(undef,L)
        ch2 = Array{Bool}(undef,L)
        for l in 1:L
            if free_recomb[l]
                ch1[l] = g[p1,l,p1ch]
                ch2[l] = g[p2,l,p2ch]
            else
                ch1[l] = g[p1,l,p2ch]
                ch2[l] = g[p2,l,p1ch]
            end
        end

        gâ‚š[i,:,1] = ch1
        gâ‚š[i,:,2] = ch2

        # poisson distr abundances centered on mid-parent for each taxa
        mLâ‚˜ = 0.5 .* (mL[p1,:] .+ mL[p2,:])
        mLâ‚š[i,:] = rand.(Poisson.(mLâ‚˜))

    end

    # offspring allele count matrix
    câ‚š = gâ‚š[:,:,1] .+ gâ‚š[:,:,2]

    # offspring environmental microbiome matrix
    mEâ‚š = rand(Poisson(KE),n,SE)

    #
    # offspring trait components
    #

    # additive genetic value
    gAâ‚š = câ‚š*Î³

    # lineal microbial value
    mlâ‚š = mLâ‚š*Ï‰L

    # nonlineal microbial value
    mnâ‚š = mNâ‚š*Ï‰N

    # environmental microbial value
    meâ‚š = mEâ‚š*Ï‰E

    # additive microbial value
    mAâ‚š = mlâ‚š .+ mnâ‚š .+ meâ‚š

    # expected developmental noise / environmental deviation
    # defined for parental pop, but stays same here for consistency
    ğ”¼E = (1 - hÂ² - Î»Â² - Î½Â² - ÏµÂ²)*ğ”¼P

    # offspring dev noise / env dev
    eâ‚š = rand(Normal(0,âˆšğ”¼E),n)

    # compute offspring trait values
    zâ‚š = gAâ‚š .+ mAâ‚š .+ eâ‚š

    # offspring data
    pdâ‚š = PD(
        g = gâ‚š,
        mL = mLâ‚š,
        mN = mNâ‚š,
        mE = mEâ‚š,
        z = zâ‚š
    )

    # offspring mean trait
    zÌ„â‚š = mean(zâ‚š)

    return zÌ„â‚š, pdâ‚š

end

function timeseries(par::Par,pp::PP,pdâ‚€::PD,T::Int64)
    @unpack_PP pp
    @unpack_PD pdâ‚€

    pdâ‚œ = Vector{PD}(undef, T)

    pdâ‚œ[1] = pdâ‚€

    for t in 1:(T-1)

        pairs, W, ğ”¼W = selection(par, pdâ‚œ[t])

        zÌ„â‚š, pdâ‚œ[t+1] = offspring(par, pp, pdâ‚œ[t], pairs)

    end

    return pdâ‚œ

end

function runs(par::Par)
    @unpack_Par par

    pp = Vector{PP}(undef,nr)
    pd = Vector{PD}(undef,nr)

    Î¼Î”zÌ„ = zeros(nr)
    Î¼Aáµ§Î² = zeros(nr)
    Î¼Aâ‚—Î² = zeros(nr)
    Î¼Aâ‚™Î² = zeros(nr)
    Î¼Aáµ©Î² = zeros(nr)

    Aáµ§ = zeros(nr)
    Aâ‚— = zeros(nr)
    Aâ‚™ = zeros(nr)
    Aáµ© = zeros(nr)

    Î”zÌ„ = zeros(sr)
    Aáµ§Î² = zeros(sr)
    Aâ‚—Î² = zeros(sr)
    Aâ‚™Î² = zeros(sr)
    Aáµ©Î² = zeros(sr)

    for i in 1:nr

        pp[i], pd[i] = init(par)

        Aáµ§[i] = GA(pp[i],pd[i])

        Aâ‚—[i] = LA(pp[i],pd[i])

        Aâ‚™[i] = TA(pp[i],pd[i])

        Aáµ©[i] = AA(pp[i],pd[i])

        zÌ„ = mean(pd[i].z)

        P = var(pd[i].z)
        
        for j in 1:sr

            pairs, W, ğ”¼W = selection(par, pd[i])

            zÌ„â‚š = zeros(or)
            for k in 1:or
                zÌ„â‚š[k], pdâ‚š = offspring(par, pp[i], pd[i], pairs)
            end

            Î”zÌ„[j] = mean(zÌ„â‚š) - zÌ„

            Î² = cov(W,pd[i].z) / P

            Aáµ§Î²[j] = Aáµ§[i]*Î²

            Aâ‚—Î²[j] = Aâ‚—[i]*Î²

            Aâ‚™Î²[j] = Aâ‚™[i]*Î²

            Aáµ©Î²[j] = Aáµ©[i]*Î²

        end

        Î¼Î”zÌ„[i] = mean(Î”zÌ„)
        Î¼Aáµ§Î²[i] = mean(Aáµ§Î²)
        Î¼Aâ‚—Î²[i] = mean(Aâ‚—Î²)
        Î¼Aâ‚™Î²[i] = mean(Aâ‚™Î²)
        Î¼Aáµ©Î²[i] = mean(Aáµ©Î²)

    end

    return Î¼Î”zÌ„, Î¼Aáµ§Î², Î¼Aâ‚—Î², Î¼Aâ‚™Î², Î¼Aáµ©Î²

end
